### DISCUSSION TITLE
Save in db as -1 day to make it possible compare our own measurements

### DISCUSSION THREAD

**Discussion by vahaldor** [DarwinsBuddy/WienerNetzeSmartmeter#195] [Ideas]
Wiener Netze reports data that are from yesterday consumption, which denies the possibility to compare our own measurements with reported data. Actually what Netze reports is saved as "today", while it reports what happened yesterday. There are of course possible methods to solve the problem later but saving the data in this component with timestamp of a day during which the consumption really happened would be nice. Or is there some solution the others followed?

**Comment by reox** [comment]
That would be really nice, but without crude hacks in home assistant (i.e., directly writing into the database), it seems not to be possible right now. The problem is, that states are written into to the database with the timestamp of data creation, rather than with an arbitrary timestamp.

Maybe something has changed in the API since I looked into this, thus if you have any useful information how this could be implemeted, let us know!

**Reply by ckadluba** [reply]
Here is a post from someone with the same problem who created an SQL script to achieve that. But as you said, this is not a quite clean way to do it.

https://community.home-assistant.io/t/how-to-import-historical-energy-data-toon-and-others/556356

**Reply by ckadluba** [reply]
... and here is a post claiming that HA has an API supporting the import of historical data. It looks worth giving it a try.

https://community.home-assistant.io/t/import-old-energy-readings-for-use-in-energy-dashboard/341406/9

**Reply by reox** [reply]
Yes, this is how it is implemented for the statistics sensor. However, if I'm not mistaken, note that these are _statistics_ and not _states_! These are two very different concepts. 

For the SQL method: I would not want to implement this in a addon. The database layout might simply change and then you have to rewrite everything. But you can of course give it a try and check if it works.

**Reply by ckadluba** [reply]
I absolutely agree on the SQL part. It's a hacky workaround which invites future problems.

Regarding the statistics sensor of the WNSM integration, do I need to activate or configure it explicitly? Because in my environment the statistics sensor never produced any data.

**Reply by reox** [reply]
No, it should be activated by default - but you are right: it does not have a state but only statistics. If you use this sensor in the energy dashboard, you can retrieve statistics information from it. See also the FAQ #19

**Reply by ckadluba** [reply]
Sorry, i was wrong. The _statistics sensor has data on my HA but it is also one day off.

But one more question. The post I linked above mentions that historical data can be imported in HA using `async_external_statistics` and you said the WNSM integration uses that. Maybe I missed it but I could not find a reference to that function in the source code.

**Reply by reox** [reply]
We use `homeassistant.components.recorder.statistics.async_import_statistics`. I once looked up what the difference is to external_statistics but cannot remember...

**Reply by ckadluba** [reply]
From what I read statistics is based on an existing sensor in HA and external statistics is not based on a sensor but comes from an external source. But for the 1 day delay problem this seems not relevant.

Also I found this thread https://community.home-assistant.io/t/improved-support-for-long-term-historic-data/379659 supporting the assumption that historic data import is just not supported in HA up to this day. :(

**Comment by mmikitz** [comment]
I am in the same situation. I have a bunch of smart plugs measuring some appliances and find myself regularly comparing how their sum differs from what the smart meter reports. Having the difference calculated as a template sensor added to the energy dashboard would allow me to spot untracked consumers. 

Any ideas how to get this implemented?

**Comment by DarwinsBuddy** [comment]
Since [v1.7.0](https://github.com/DarwinsBuddy/WienerNetzeSmartmeter/releases/tag/v1.7.0) we've got only one sensor left that we maintain. The other one was a bit of a workaround.
See [v1.7.0 what changed?](https://github.com/DarwinsBuddy/WienerNetzeSmartmeter/wiki/FAQs#since-v170-why-is-my-_statistics-sensor-not-updated-any-longer-with-historical-data) for more details.
We managed to get the factually reported data of your smartmeters by pushing them into the statistics (recorder) module of home assistant. If you want to have reliable data please use those. i.e. `wnsm:at10000000000000000010000000` instead of the real sensor entity. The latter merely tracks the meter readings at a specific point in time when it can be scraped. It's not suitable for time-relevant comparison, due to the delay added by the roundtrip smartmeter -> WienerNetze -> Your Home Assistant and the fact that the sensor entity cannot be used in that sense.
